jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
@akosthekiss Heh, during our design discussions, me and @jiangzidong had a very similar sketch at some point as what you just posted.

> Passing that struct to the transforms exposes too much of the internals of the validate-and-assign algorithm.
> All they should care about what the JS argument is and where to store the C value (if possible).
What the internal pointers and counters are should not be of their concern.

I agree the iterator should be internal. I think the solution is just to move the `typedef jerryx_arg_iterator_t` into the .c file and make the one in the .h a forward type declaration.
Note that the transforms do not access the internals of the `jerryx_arg_iterator_t` directly, but only through the iterator accessor functions.

Personally, I'd much rather pass a single iterator pointer that encapsulates the state instead of multiple arguments of which some will not be used most of the time (`js_arg_cnt` for example is only relevant to transforms that process multiple `jerry_value_t`s).
More importantly, if we want to enable more complex behavior (as future API or for customization inside a project), it's easy to add more iterator accessors (no API breakage), but it's not easy to add arguments to the transformer function signature because everything would have to be changed (causing API breakage).

[example]: https://github.com/jerryscript-project/jerryscript/pull/1740#issuecomment-297300966